### PR TITLE
amp-accordion: add `store-state-opt-out` attribute to disable store state

### PR DESCRIPTION
**Pull request by zhouyx**
#3813. Needed because of #3298

Put this feature into experiment before validator supports the attribute.
Also please suggest better attribute name.

### CODE REVIEW

**Comment by ericlindley-g** [comment]
@zhouyx @sriramkrish85 

What do you think about "reset-state", "disable-state-storage", or "reset-state-on-load"?

**Comment by ericlindley-g** [comment]
Also, a quick note that we'll want to document this in the .md as well, when it's live

File: extensions/amp-accordion/0.1/amp-accordion.js [hunk 1]
```diff
@@ -28,22 +32,34 @@ class AmpAccordion extends AMP.BaseElement {
...
 
-    /** @const @private {string} */
-    this.id_ = this.getSessionStorageKey_();
+    /** @const @private {string|null} */
+    this.sessionId_ = null;
+
+    /** @const @private {boolean} */
+    this.sessionOptOut_ = false;
+    if (this.isExperimentOn_) {
+      this.sessionOptOut_ = this.element.hasAttribute('store-state-opt-out');
```
Comment (sriramkrish85) [inline extensions/amp-accordion/0.1/amp-accordion.js]: i am not convinced with this name. 
`disable-session-states`
`incognito` ?

@cramforce wdyt?

File: extensions/amp-accordion/0.1/amp-accordion.js [hunk 2]
```diff
@@ -15,10 +15,14 @@
  */
 
 import {CSS} from '../../../build/amp-accordion-0.1.css';
+import {isExperimentOn} from '../../../src/experiments';
 import {Layout} from '../../../src/layout';
 import {dev, user} from '../../../src/log';
 import {removeFragment} from '../../../src/url';
 
+/** @const */
+const TAG = 'amp-accordion-store-state-opt-out';
```
Comment (sriramkrish85) [inline extensions/amp-accordion/0.1/amp-accordion.js]: amp-accordion-session-store-optout ?

File: extensions/amp-accordion/0.1/amp-accordion.js (hunk 1 above)
Comment (ericlindley-g) [inline extensions/amp-accordion/0.1/amp-accordion.js]: Other potential options:
"reset-state" or "reset-state-on-load" ?

**Comment by cramforce** [comment]
Did we determine that this is needed, given that the storage is per URL and per window?

**Comment by sriramkrish85** [comment]
@zhouyx what happened here? can you update

**Comment by ericlindley-g** [comment]
I think we questioned that this was the right approach to solve the problem. @cramforce: do you have any reservations? On reviewing the comment thread, the original issue seems like it would be solved by this change.

**Comment by zhouyx** [comment]
The current approach disable storage per `amp-accordion` element. I don't have a good way to disable/enable storage per URL. 
If we want to do this per `amp-accordion` I guess the left thing would to choose a proper name.

**Comment by ericlindley-g** [comment]
I had been assuming per-element opt-out, so that sounds good to me. Does anyone else have opinions on that or the name?

**Comment by sriramkrish85** [comment]
SGTM

**Comment by zhouyx** [comment]
Current naming options:
1. `disable-session-states`
2. `incognito`
3. `reset-state`
4. `reset-state-on-load`

**Comment by zhouyx** [comment]
I will use the name `disable-session-states` for the attribute, and `amp-accordion-session-state-optout` for the experiment name. If we are good with the name then let's preceed with this
PTAL

File: extensions/amp-accordion/0.1/amp-accordion.js [hunk 3]
```diff
@@ -28,22 +32,34 @@ class AmpAccordion extends AMP.BaseElement {
...
 
-    /** @const @private {string} */
-    this.id_ = this.getSessionStorageKey_();
+    /** @const @private {string|null} */
+    this.sessionId_ = null;
+
+    /** @const @private {boolean} */
+    this.sessionOptOut_ = false;
+    if (this.isExperimentOn_) {
+      this.sessionOptOut_ = this.element.hasAttribute('disable-session-states');
```
Comment (cramforce) [inline extensions/amp-accordion/0.1/amp-accordion.js]: Any reason this would be plural?

File: extensions/amp-accordion/0.1/amp-accordion.js [hunk 4]
```diff
@@ -28,22 +32,34 @@ class AmpAccordion extends AMP.BaseElement {
 
   /** @override */
   buildCallback() {
+    /** @const @private {boolean} */
+    this.isExperimentOn_ = isExperimentOn(this.getWin(), TAG);
```
Comment (cramforce) [inline extensions/amp-accordion/0.1/amp-accordion.js]: This is a specific experiment.
I missed previous discussion, but I don't think we need a feature flag for this.

File: extensions/amp-accordion/0.1/amp-accordion.js (hunk 3 above)
Comment (zhouyx) [inline extensions/amp-accordion/0.1/amp-accordion.js]: No specific reason to me, we can always change to `disable-session-state`

File: extensions/amp-accordion/0.1/amp-accordion.js (hunk 4 above)
Comment (zhouyx) [inline extensions/amp-accordion/0.1/amp-accordion.js]: It's because we want to guard this property before validator provide support for this attribute.

**Comment by camelburrito** [comment]
Lets scrap this PR - since we have had this for a while - this should be the  new norm? @ericlindley-g WDYT

**Comment by ericlindley-g** [comment]
Is there more work to be done on this, or a high maintenance cost? There are still users that want the feature, and I think the only remaining work is on the validator. It's not high-priority, but if it's nearly done, I'm inclined just to finish it up and merge, as opposed to scrapping it.

**Comment by zhouyx** [comment]
wow, this has been here for a long time. @cramforce Did you remember your concern about this PR? I think @camelburrito approved it.

**Comment by jridgewell** [comment]
@zhouyx, we have another request for this in #6194.

File: extensions/amp-accordion/0.1/amp-accordion.js [hunk 5]
```diff
@@ -44,13 +51,20 @@ class AmpAccordion extends AMP.BaseElement {
...
+    }
 
     this.element.setAttribute('role', 'tablist');
 
-    // sessionStorage key: special created id for this element, this.id_.
+    // sessionStorage key: special created id for this element, this.sessionId_.
     // sessionStorage value: string that can convert to this.currentState_ obj.
-    this.currentState_ = this.getSessionState_();
+    if (!this.sessionOptOut_) {
+      this.sessionId_ = this.getSessionStorageKey_();
```
Comment (jridgewell) [inline extensions/amp-accordion/0.1/amp-accordion.js]: duplicated

File: extensions/amp-accordion/0.1/amp-accordion.js [hunk 6]
```diff
@@ -44,13 +51,20 @@ class AmpAccordion extends AMP.BaseElement {
...
+      this.sessionOptOut_ = this.element.hasAttribute('disable-session-states');
+    }
 
     this.element.setAttribute('role', 'tablist');
 
-    // sessionStorage key: special created id for this element, this.id_.
+    // sessionStorage key: special created id for this element, this.sessionId_.
     // sessionStorage value: string that can convert to this.currentState_ obj.
-    this.currentState_ = this.getSessionState_();
+    if (!this.sessionOptOut_) {
```
Comment (jridgewell) [inline extensions/amp-accordion/0.1/amp-accordion.js]: This can be consolidated into `#getSessionState` and `#setSessionState`.

Comment (zhouyx) [inline extensions/amp-accordion/0.1/amp-accordion.js]: done.
Cons: we will generate sessionStorageKey for optout amp-accordion even it is not needed, Also I will keep record of current state during runtime even opt out. (but not write to session).
But the consolidation did make code cleaner, I suppose it is fine.

File: extensions/amp-accordion/0.1/amp-accordion.js (hunk 5 above)
Comment (zhouyx) [inline extensions/amp-accordion/0.1/amp-accordion.js]: deleted

**Comment by zhouyx** [comment]
PTAL

**Comment by camelburrito** [comment]
@zhouyx - please test if the component works as intended :)

**Comment by zhouyx** [comment]
@camelburrito Already tested

**Pull request merged**